plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -1,915 +1,1218 @@
+// are supported package-specific, and there are no duplicates. If metrics slice is nil, then
+// default package metrics are provided.
+func parsePackageMetrics(metrics []string) ([]string, error) {
+	if metrics == nil {
+		return []string{
+			packageCurrentPowerConsumption.String(),
+			packageCurrentDramPowerConsumption.String(),
+			packageThermalDesignPower.String(),
+		}, nil
+	}
+
+	for _, m := range metrics {
+		if !isValidPackageMetric(m) {
+			return nil, fmt.Errorf("invalid package metric specified: %q", m)
 		}
 	}
-	initMinFreq, err := p.msr.retrieveUncoreFrequency(socketID, typeFreq, "min", die)
-	if err != nil {
-		p.Log.Errorf("Error while retrieving minimum uncore frequency of the socket %s, err: %v", socketID, err)
-		return
-	}
-	initMaxFreq, err := p.msr.retrieveUncoreFrequency(socketID, typeFreq, "max", die)
-	if err != nil {
-		p.Log.Errorf("Error while retrieving maximum uncore frequency of the socket %s, err: %v", socketID, err)
-		return
-	}
 
-	tags := map[string]string{
-		"package_id": socketID,
-		"type":       typeFreq,
-		"die":        die,
+	if hasDuplicate(metrics) {
+		return nil, errors.New("package metrics contains duplicates")
 	}
-	fields["uncore_frequency_limit_mhz_min"] = initMinFreq
-	fields["uncore_frequency_limit_mhz_max"] = initMaxFreq
-
-	acc.AddGauge("powerstat_package", fields, tags)
+	return metrics, nil
 }
 
-func (p *PowerStat) addThermalDesignPowerMetric(socketID string, acc telegraf.Accumulator) {
-	maxPower, err := p.rapl.getConstraintMaxPowerWatts(socketID)
-	if err != nil {
-		p.Log.Errorf("Error while retrieving TDP of the socket %s, err: %v", socketID, err)
-		return
+// parseCPUMetrics takes a slice of CPU metrics. It ensures all metrics in the slice are
+// supported CPU-specific, and there are no duplicates.
+func parseCPUMetrics(metrics []string) ([]string, error) {
+	for _, m := range metrics {
+		if !isValidCoreMetric(m) {
+			return nil, fmt.Errorf("invalid core metric specified: %q", m)
+		}
 	}
 
-	tags := map[string]string{
-		"package_id": socketID,
+	if hasDuplicate(metrics) {
+		return nil, errors.New("core metrics contains duplicates")
 	}
+	return metrics, nil
+}
 
-	fields := map[string]interface{}{
-		"thermal_design_power_watts": roundFloatToNearestTwoDecimalPlaces(maxPower),
+// parsedCPUTimedMsrMetrics takes a slice of unique CPU metrics, and returns only the metrics which
+// depend on time-related MSR offset reads.
+func parseCPUTimeRelatedMsrMetrics(metrics []string) []string {
+	slice := []string{
+		cpuC0StateResidency.String(),
+		cpuC1StateResidency.String(),
+		cpuC3StateResidency.String(),
+		cpuC6StateResidency.String(),
+		cpuC7StateResidency.String(),
+		cpuBusyCycles.String(),
+		cpuBusyFrequency.String(),
+	}
+
+	cpuTimeRelatedMsrMetrics := make([]string, 0, len(slice))
+	for _, m := range metrics {
+		if choice.Contains(m, slice) {
+			cpuTimeRelatedMsrMetrics = append(cpuTimeRelatedMsrMetrics, m)
+		}
 	}
```

Comment:
Wouldn't a switch-statement be more efficient and easier to read?

Reply:
Yes, absolutely. Done as part of commit [e2de6f6b](https://github.com/influxdata/telegraf/pull/14363/commits/e2de6f6b2577fe4e9ec7b420132b2750fe7e2654).